docs/madr/decisions/049-meshexternalservice-api.md

```diff
@@ -0,0 +1,474 @@
+      * one of `inline`, `inlineString` or `secret`.
+    * **clientKey**: defines a client private key.
+      * one of `inline`, `inlineString` or `secret`.
+* **status**: status of an object managed by Kuma control-plane
+  * **addresses**:
+    * **hostname**: domain generated by HostnameGenerator
+    * **status**: if a domain is already assigned, possible values `Available` | `NotAvailable`
+    * **origin**: provided an information how the domain was generated
+      * **kind**: kind of resource responsible for hostname generation
+      * **name**: name of the resource
+    * **reason**: set when there was error when generating a hostname
+  * **vips**: list of allocated VIPs
+    * **ip**: allocated IP for a provided domain with `InternalVIP` type in a match section
+    * **type**: provides information about the way IP was provided
+    * **hostname**: provides a domain with `InternalVIP` type in a match section for which IP was allocated. In case of many entries it helps corelating entries.
+
+##### Hostname and VIP generation
+
+Currently, Kuma allocates a domain for an ExternalService based on the kuma.io/service label and the real destination domain. However, in MeshExternalService, we aim to discontinue this practice. Instead, we propose utilizing the HostGenerator for domain generation. To facilitate this change, we intend to introduce a new selector called meshExternalServiceSelector, enabling users to specify particular MeshExternalServices that necessitate custom domains. Only MeshOperator should create a HostGenerator policy and when a user needs a specific domain for a MeshExternalService, the user needs to reach out to the MeshOperator. HostGenerator can be created only on GlobalCP and later is synchronized to all zones. More about HostGenerator (MADR-046).
+
+```yaml
+kind: HostnameGenerator
+metadata:
+  name: k8s-zone-hostnames
+spec:
+  targetRef:
+    kind: MeshExternalService
+    tags:
+      my.label.io/service: my-service 
+  template: {{ name }}.svc.meshext.local
+```
+
+For example, given this `MeshExternalService`
+
+```yaml
+kind: MeshExternalService
+metadata:
+  name: mydomain
+  namespace: kuma-system
+  labels:
+    my.label.io/service: my-service 
+    kuma.io/mesh: default
+    kuma.io/zone: east-1
+    kuma.io/origin: zone
+spec:
+  destinations:
+  - address: 192.168.0.1
+    port: 9090 
+...
+```
+
+We would generate such hostname `mydomain.svc.meshext.local`
+
+Possible template functions/keys are:
+
+`{{ name }}` - name of the `MeshExternalService`
+`{{ label "x" }}` - value of label x.
+
+If the template cannot be resolved (label is missing), the hostname won't be generated with a NotAvailable status and an appropriate reason.
+
+#### Positive Consequences
+
+* Simpler model, but not as clear
+* One domain per external service 
+
+#### Negative Consequences
+
+* Using HostnameGenerator might cause unexpected domains to be created, when match is too wide
```

Comment:
I think I'd add here that matching to a hostname generated by HostnameGenerator in `match` section looks a bit awkward.